File: tyhp-lang/vscode/src/debug/waitForTcp.ts
import * as net from "node:net";

/** True when `host:port` accepts a TCP connection within `timeoutMs`. */
export function probeTcpPort(host: string, port: number, timeoutMs: number): Promise<boolean> {
    if (port <= 0) {
        return Promise.resolve(false);
    }
    return new Promise((resolve) => {
        const socket = net.connect({ host, port });
        const finish = (ok: boolean) => {
            socket.removeAllListeners();
            socket.destroy();
            resolve(ok);
        };
        socket.setTimeout(timeoutMs);
        socket.once("connect", () => finish(true));
        socket.once("timeout", () => finish(false));
        socket.once("error", () => finish(false));
    });
}

export function sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

File: tyhp-lang/phpstorm/src/main/kotlin/com/tyhp/lang/debug/WaitForTcp.kt
package com.tyhp.lang.debug

import java.net.InetSocketAddress
import java.net.Socket

const val PROXY_LISTEN_ADDRESS = "127.0.0.1"

/** True when `host:port` accepts a TCP connection within [timeoutMs]. */
fun probeTcpPort(host: String, port: Int, timeoutMs: Int): Boolean {
    if (port <= 0) {
        return false
    }
    Socket().use { socket ->
        return try {
            socket.connect(InetSocketAddress(host, port), timeoutMs)
            true
        } catch (_: Exception) {
            false
        }
    }
}

fun sleepMs(ms: Long) {
    Thread.sleep(ms)
}
